analysis/import/SCMLogParserV2/src/main/kotlin/de/maibornwolff/codecharta/importer/scmlogparserv2/parser/VersionControlledFilesList.kt

```diff
@@ -0,0 +1,119 @@
+package de.maibornwolff.codecharta.importer.scmlogparserv2.parser
+
+import de.maibornwolff.codecharta.importer.scmlogparserv2.input.VersionControlledFile
+import de.maibornwolff.codecharta.importer.scmlogparserv2.input.metrics.MetricsFactory
+
+class VersionControlledFilesList(private val metricsFactory: MetricsFactory) {
+
+    private var versionControlledFiles: MutableMap<String, VersionControlledFile> = mutableMapOf()
+
+    /**
+     * The key of the map is the current name
+     * The value of the map is the oldest name
+     */
+    private var renamesMap: MutableMap<String, String> = mutableMapOf()
+
+    /**
+     * The key of the map is the current name
+     * The value of the map is the number of name conflicts as a counter
+     */
+    private var nameConflictsMap: MutableMap<String, Int> = mutableMapOf()
+
+    fun get(key: String): VersionControlledFile? {
+        return versionControlledFiles[resolveFileKey(key)]
+    }
+
+    fun addFileBy(key: String): VersionControlledFile {
+        var newVCFFileName = key
+
+        if (hasNameConflict(key)) {
+            newVCFFileName = handleNameConflict(key)
+        }
+
+        val vcf = VersionControlledFile(newVCFFileName, metricsFactory)
+
+        versionControlledFiles[resolveFileKey(key)] = vcf
+        return vcf
+    }
+
+    private fun hasNameConflict(key: String): Boolean {
+        return versionControlledFiles.containsKey(key) && !versionControlledFiles[key]!!.isDeleted()
+    }
+
+    private fun handleNameConflict(key: String): String {
+        val marker = nameConflictsMap[key]
+        val newMarker = if (marker != null) marker + 1 else 0
+        nameConflictsMap[key] = newMarker
+
+        return key + "_\\0_" + newMarker
+    }
+
+    fun getList(): MutableMap<String, VersionControlledFile> {
+        return versionControlledFiles
+    }
+
+    fun rename(oldFileName: String, newFileName: String) {
+        var newVCFFileName = newFileName
+        val possibleConflictName = buildPossibleConflictName(oldFileName)
+        val oldestName = retrieveOldestName(possibleConflictName)
+
+        if (fileExistsAsDeleted(newFileName)) {
+            handleDeletedFileReplacedByRenamedFile(newFileName)
+        }
+
+        if (versionControlledFiles.containsKey(newFileName) && !isCyclicRename(oldFileName, newFileName)) {
+            newVCFFileName = handleNameConflict(newFileName)
+        }
+
+        if (oldestName != null) {
+            renamesMap.remove(possibleConflictName)
+        }
+
+        val updatedOldestName = if (oldestName != null) oldestName else oldFileName
+        renamesMap[newVCFFileName] = updatedOldestName
+
+        val notYetRenamedFile = versionControlledFiles[updatedOldestName]!!
+        notYetRenamedFile.addRename(newVCFFileName)
+        notYetRenamedFile.filename = newFileName
+    }
+
+    private fun isCyclicRename(oldFileName: String, newFileName: String): Boolean {
+        return get(oldFileName)!!.containsRename(newFileName)
```

Comment:
Add comment why this is necessary.

Reply:
You had a nice explanation yesterday, would you mind repeating it?